### ISSUE TITLE
Getters/setters that use get_node need a manual guard.

### ISSUE THREAD

**Issue by QbieShay** [godotengine/godot#28014]
**Godot version:**
3.1-stable

**Issue description:**
When using getters and setters, there might be cases where you want to access child node properties (this is expecially true when you need to "forward" certain properties higher up in the hierarchy).
The current solution to this is to manually put a `has_node` in front of the `get_node("name").stuff = stuff` and then setting manually those properties in the _ready() function.
Seems like this problem presents itself only coupled with tool scripts

**Proposal**:
setters are called only after the full node hierarchy has been succesfully constructed (possibly starting from the leaf nodes)

**Steps to reproduce:**
1. download mrp
2. start the game - works
3. add `tool` on the BallFw.gd script
4. start the game - observe an error on start
From now on the game will always be unable to properly start, even when closing and reopening the engine.

**Minimal reproduction project:**
[setter_crash.zip](https://github.com/godotengine/godot/files/3077403/setter_crash.zip)

**Comment by Zylann** [comment]
> setters are called only after the full node hierarchy has been succesfully constructed (possibly starting from the leaf nodes)

This will break compatibility and the case where child nodes need to access the state of their parent.
I think if a node needs to access a child in its setter, that suggests a strong dependency to this child, and that child would be added in `_init` rather than later, making it accessible even before entering the tree (the same way composite C++ nodes like `ColorPicker` create their children in their constructor too).

**Comment by QbieShay** [comment]
> On the other hand, if the node you try to access can be removed or renamed by a designer, it's not a good idea to assume it's always there anyways. If nothing in code enforces or checks that, it opens the door for errors.

Being visible in the editor doesn't necessarily mean that a designer is going to edit it. For example, having 3D objects editable in the editor is still good if you want to tweak properties such as materials, shapes, transforms, etc. (otherwise we wouldn't need a visual editor).
There are many ways to get around "the designers problem" (for example if you don't want them to see some folders something similar to Unreal's "developer folders" could be implemented)

I don't have a strong preference between the order of the setters :) probably it actually makes more sense to have the setters be set on the parent nodes first.

**Comment by Jimmio92** [comment]
I encountered this problem just now and had no idea it had to do with the script being a tool until I searched here on GitHub and found this bug report. I was developing a stat bar to display stamina from my player, and get_node on the Control parent would fail, saying I was trying to call it outside of the tree with error "Can't use get_node() with absolute paths from outside the active scene tree." I thought for sure that Godot got confused, seeing as I have 4 separate Viewports with 4 separate worlds. I had to comment out the tool keyword, and totally restart Godot before I could continue developing. `is_in_tree()` was also returning true, so I couldn't even hack around it by early returning if outside of tree.

Once I got back into Godot after removing the tool keyword, it works fine, even with the tool keyword added back? Surely this is an actual bug that needs fixing. I was extremely frustrated, because as far as I could tell, this was stupidity on Godot's part, not my own, and being unable to share the project due to copyright/trademark concerns meant getting anyone to understand what was going on was going to be a challenge.

Given what Zylann just mentioned, I should be able to have it a tool still if I initialize the child node references in _init, meaning I need to start *always* getting references needed in _init.

Sorry for the bit of a ramble, but THANK YOU for helping me get around this oddball bug.
